sklearn/metrics/regression.py

```diff
@@ -684,47 +683,8 @@ def mean_tweedie_deviance(y_true, y_pred, sample_weight=None, power=0):
         sample_weight = column_or_1d(sample_weight)
         sample_weight = sample_weight[:, np.newaxis]
 
-    message = ("Mean Tweedie deviance error with power={} can only be used on "
-               .format(power))
-    if power < 0:
-        # 'Extreme stable', y_true any realy number, y_pred > 0
-        if (y_pred <= 0).any():
-            raise ValueError(message + "strictly positive y_pred.")
-        dev = 2 * (np.power(np.maximum(y_true, 0), 2 - power)
-                   / ((1 - power) * (2 - power))
-                   - y_true * np.power(y_pred, 1 - power)/(1 - power)
-                   + np.power(y_pred, 2 - power)/(2 - power))
-    elif power == 0:
-        # Normal distribution, y_true and y_pred any real number
-        dev = (y_true - y_pred)**2
-    elif power < 1:
-        raise ValueError("Tweedie deviance is only defined for power<=0 and "
-                         "power>=1.")
-    elif power == 1:
-        # Poisson distribution, y_true >= 0, y_pred > 0
-        if (y_true < 0).any() or (y_pred <= 0).any():
-            raise ValueError(message + "non-negative y_true and strictly "
-                             "positive y_pred.")
-        dev = 2 * (xlogy(y_true, y_true/y_pred) - y_true + y_pred)
-    elif power == 2:
-        # Gamma distribution, y_true and y_pred > 0
-        if (y_true <= 0).any() or (y_pred <= 0).any():
-            raise ValueError(message + "strictly positive y_true and y_pred.")
-        dev = 2 * (np.log(y_pred/y_true) + y_true/y_pred - 1)
-    else:
-        if power < 2:
-            # 1 < p < 2 is Compound Poisson, y_true >= 0, y_pred > 0
-            if (y_true < 0).any() or (y_pred <= 0).any():
-                raise ValueError(message + "non-negative y_true and strictly "
-                                           "positive y_pred.")
-        else:
-            if (y_true <= 0).any() or (y_pred <= 0).any():
-                raise ValueError(message + "strictly positive y_true and "
-                                           "y_pred.")
-
-        dev = 2 * (np.power(y_true, 2 - power)/((1 - power) * (2 - power))
-                   - y_true * np.power(y_pred, 1 - power)/(1 - power)
-                   + np.power(y_pred, 2 - power)/(2 - power))
+    dist = TweedieDistribution(power=power)
+    dev = dist.unit_deviance(y_true, y_pred, check_input=True)
```

Comment:
I would not have done this. It creates intricates imports between different modules of the code base which leads to circular imports. I would keep it as it is make the GLM code use metrics instead.

Reply:
The thing is that we cant use metrics instead, because `mean_tweedie_deviance` averages over samples while `TweedieDistribution.unit_deviance` that we use here doesn't. We could define unit devince for tweedie as another private function in metrics, but that's a bit ugly, as it would then be defined separately from its derivatives.

Circular imports shouldn't be an issue as it's a function level import, but module interdependence is a problem. Maybe it could be the occasion to put all losses in one place e.g. under `sklearn/_loss/` and move into the same module losses from `sklearn/ensemble/_hist_gradient_boosting/loss.py`, `sklearn/ensemble/_gb_losses.py` and those from `sklearn/linear_model/sgd_fast.pyx` to see what could be re-used.

Reply:
@rth I like your idea of one common place for loss functions very much. 👍

Reply:
Thanks for the feedback! Moved it under `sklearn._loss.glm_distributions` and opened https://github.com/scikit-learn/scikit-learn/issues/15123 for a more general discussion.